rfcs/0007-synchronous-commit.md

```diff
@@ -0,0 +1,442 @@
+
+While `SyncLevel` and `DurabilityLevel` may appear similar, they serve distinct purposes. `SyncLevel` controls when writes become visible to readers by determining the commit semantics, while `DurabilityLevel` focuses on the durability of the write.
+
+In my opinion, keeping these as separate enums follows the Single Responsibility Principle, as each enum has a distinct purpose. They are not necessarily to have to be the same thing in the future. For example, in the PostgreSQL example, the option `remote_apply` is not really about the durability of the write, but the application of the commit. We have no plan to support it in SlateDB at the time of writing, but it might be a good example to distinguish the concept difference between `SyncLevel` and `DurabilityLevel`.
+
+## Implementation
+
+The key change is to seperate "Append to WAL" and "Commit to make it visible" into two different steps.
+
+In the current design, whenever a write operation is called, it'll append the data to the WAL, the data is becoming visible to readers with `DurabilityLevel::Memory` immediately, as the readers will read the data from the WAL in front of MemTable. And when the WAL buffer reaches the `flush.interval` or `flush.size`, it'll be flushed to storage, and apply to MemTable.
+
+The sequence diagram is like this:
+
+```mermaid
+sequenceDiagram
+    participant WriteBatch
+    participant WAL
+    participant MemTable
+    participant Flusher
+
+    WriteBatch->>WAL: append write op to current WAL
+    Note over WAL: become visible to readers with DurabilityLevel::Memory
+    WAL->>WAL: maybe trigger a flush
+    Note over WAL: freeze the current WAL, and flush this WAL
+    WAL->>Flusher: Flush WAL
+    Flusher-->>WAL: ACK FlushWAL message
+    WAL->>MemTable: Merge WAL to MemTable
+```
+
+In this proposal, "Commit to make it visible" effectively means "applying the changes to the MemTable", whenever the change is applied to MemTable, it's considered as committed.
+
+The changes to the read path are minimal. Readers with `ReadWatermark::LastCommitting` will continue to read data from the WAL before accessing the MemTable. For reads with `ReadWatermark::LastCommitted`, `ReadWatermark::LastLocalPersisted`, or `ReadWatermark::LastRemotePersisted`, readers will first check the MemTable, using different sequence numbers as watermarks to determine visibility.
+
+On the write side, after the data is appended to the WAL, it'll be applied to MemTable as soon as possible to make the change to be visible to readers.
+
+The sequence diagram is like this:
+
+```mermaid
+sequenceDiagram
+    participant WriteBatch
+    participant WAL
+    participant MemTable
+    participant Flusher
+
+    WriteBatch->>WAL: append write op to current WAL
+    Note over WAL: become visible to readers with ReadWatermark::LastCommitting
```

Comment:
LastCommitting is gone now, right?

Reply:
oh yes, thank you for reminding, let me remove this line

Reply:
removed in a27a65e